### PR TITLE
PyTorch Implementation

### PR DESCRIPTION
**Pull request by 4rash-uni**
* [x] finish base model and losses
* [x] merge PredicateAlignModel to DataModel
* [x]  multiple Dataset for each view from merged data model using native PyTorch approach (some views need same dataset)
* [x]  finish all models
* [x]  training process

### CODE REVIEW

**Comment by 4rash-uni** [comment]
### TODO List
- finish MultiKELoss
- finish test, valid MultiKENet
- merge PredictAlignModel to DataModel
- multiple Dataset for each view from merged data model using native pytorch approach (some views need same dataset) TensorDataset could be used like literal_encoder
- finish all models
- training process

@KirillDogadin-std @mehmetbalci

**Comment by KirillDogadin-std** [comment]
> * [ ]  multiple Dataset for each view from merged data model using native pytorch approach (some views need same dataset) TensorDataset could be used like literal_encoder
> * [ ]  finish all models
> * [ ]  training process

These 3 are not really clear, I think the context lacks. 

Do you perhaps have an url that can explain what you mean in the first quoted bullet?

The other 2 are very general, I assume that you're going to adjust them later down the road.

**Comment by KirillDogadin-std** [comment]
Opened PRs #14  #13  #12

File: code/pytorch/data_model.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    literal_id_dic = dict()
+    print('literal id', len(literal_list), len(set(literal_list)))
+    for i in range(len(literal_list)):
+        literal_id_dic[literal_list[i]] = i
+    assert len(literal_list) == len(literal_id_dic)
+    return literal_id_dic
+
+
+class DataModel:
+    def __init__(self, args):
```
Comment (KirillDogadin-std) [inline code/pytorch/data_model.py:40]: here the sigatire is still just `args` how it was initally.

**Review by KirillDogadin-std — COMMENTED** [review]
Could you please provide what is this that you expect to see in the extented-model classes which use the merged `DataModel`? 

I observe that some function like `__init__` in `torch/data_model.py` have now a very different layout from what is `tf`.  (Or one can say a _pipeline_). My assumtion here that there's no intention to pass `data` to the `model` at all.

I would appreciate if you could just lay out function signatures without any contents and a usage-flow so that i could understand what you expect to see. Something like that applied to net class

```python

def class MultiKENet:
    def __init__(self, num_entities, num_relations, num_attributes, embed_dim, value_vectors, local_name_vectors):
        pass
    def forward(self, inputs, view):
        pass

if __name__ == '__main__':
    conv = Conv(75)
    score = conv(torch.rand(32, 75), torch.rand(32, 75), torch.rand(32, 75))
    print(conv)
    model = MultiKENet(200000, 550, 1086, 75, np.zeros((909911, 75), dtype=np.float32), np.zeros((200000, 75), dtype=np.float32))
    inputs = torch.randint(75, (32,)), torch.randint(75, (32,)), torch.randint(75, (32,))
    outputs = model(inputs, 'av')
    
    losses = MultiKELoss(eye, cv_name_weight, cv_weight, orthogonal_weight)

```

**Comment by 4rash-uni** [comment]
> These 3 are not really clear, I think the context lacks.
> 
> Do you perhaps have an url that can explain what you mean in the **first** quoted bullet?
> 
> The other 2 are very general, I assume that you're going to adjust them later down the road.

Since all steps is done for the ITC now, you can follow the code to find out.

**Comment by 4rash-uni** [comment]
> Could you please provide what is this that you expect to see in the extented-model classes which use the merged `DataModel`?
> 
> I observe that some function like `__init__` in `torch/data_model.py` have now a very different layout from what is `tf`. (Or one can say a _pipeline_). My assumtion here that there's no intention to pass `data` to the `model` at all.
> 
> I would appreciate if you could just lay out function signatures without any contents and a usage-flow so that i could understand what you expect to see. Something like that applied to net class




> > * [ ]   multiple Dataset for each view from merged data model using native pytorch approach (some views need same dataset) TensorDataset could be used like literal_encoder
> > * [ ]   finish all models
> > * [ ]   training process
> 
> These 3 are not really clear, I think the context lacks.
> 
> Do you perhaps have an url that can explain what you mean in the **first** quoted bullet?
> 
> The other 2 are very general, I assume that you're going to adjust them later down the road.

**Comment by 4rash-uni** [comment]
> These 3 are not really clear, I think the context lacks.
> 
> Do you perhaps have an url that can explain what you mean in the **first** quoted bullet?
> 
> The other 2 are very general, I assume that you're going to adjust them later down the road.

Since all steps are done for the ITC now, you can follow the code to find out.

**Comment by 4rash-uni** [comment]
> Could you please provide what is this that you expect to see in the extented-model classes which use the merged `DataModel`?
> 
> I observe that some function like `__init__` in `torch/data_model.py` have now a very different layout from what is `tf`. (Or one can say a _pipeline_). My assumtion here that there's no intention to pass `data` to the `model` at all.
> 
> I would appreciate if you could just lay out function signatures without any contents and a usage-flow so that i could understand what you expect to see. Something like that applied to net class

As you can see in the base model signature, only local_name_vectors and value_vectors from data model are used to create the model. The merged data model is mainly used by datasets to provide data for training and feed the provided data to the model during training.
The extended-model classes will be removed at the end since the base model is enough.

**Pull request merged**